File: dart/day10_part1_2020.dart

import 'dart:io';

void main() {
  List<int> adapters = File('input.txt').readAsLinesSync().map(int.parse).toList();
  adapters.sort();

  int oneJoltDiff = 0;
  int threeJoltDiff = 1;
  int currentJoltage = 0;

  for (int adapter in adapters) {
    if (adapter - currentJoltage == 1) {
      oneJoltDiff++;
    } else if (adapter - currentJoltage == 3) {
      threeJoltDiff++;
    }
    currentJoltage = adapter;
  }

  print(oneJoltDiff * threeJoltDiff);
}

File: day10_part1_2020.rs

use std::fs;

fn main() {
    let mut input: Vec<i32> = fs::read_to_string("input.txt")
        .expect("Unable to read file")
        .lines()
        .map(|line| line.parse().unwrap())
        .collect();

    input.sort();
    let mut one_jolt_diff = 0;
    let mut three_jolt_diff = 1; // Start with 1 for the built-in adapter

    let mut prev = 0;
    for &num in &input {
        match num - prev {
            1 => one_jolt_diff += 1,
            3 => three_jolt_diff += 1,
            _ => (),
        }
        prev = num;
    }

    println!("{}", one_jolt_diff * three_jolt_diff);
}
